azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+        self.subscription_map = {}
+        self.unsubscription_map = {}
+
+    def _run_op(self, op):
+
+        if op.name == pipeline_ops.names.IotHub_SetAuthProviderArgs:
+            # if we get auth provider args from above, we save some, use some to build topic names,
+            # and always pass it down because we know that the Mqtt_Provider stage will also want
+            # to receive these args.
+            self._set_topic_names(device_id=op.device_id, module_id=op.module_id)
+            client_id = op.device_id
+            if op.module_id:
+                client_id += "/" + op.module_id
+            username = op.hostname + "/" + client_id + "/" + "?api-version=2018-06-30"
+            if op.gateway_hostname:
+                hostname = op.gateway_hostname
+            else:
+                hostname = op.hostname
+
+            self.continue_with_translated_op(
+                original_op=op,
+                translated_op=pipeline_ops.Mqtt_SetConnectionArgs(
+                    client_id=client_id,
+                    hostname=hostname,
+                    username=username,
+                    device_id=op.device_id,
+                    module_id=op.module_id,
+                    ca_cert=op.ca_cert,
+                ),
+            )
+
+        elif op.name in (
+            pipeline_ops.names.IotHub_SendTelemetry,
+            pipeline_ops.names.IotHub_SendOutputEvent,
+        ):
+            # Convert IotHub_SendTelementry and IotHub_SendOutputEvent operations into Mqtt_Publish operations
+            topic = mqtt_topic_helper.encode_properties(op.message, self.telemetry_topic)
+            self.continue_with_translated_op(
+                original_op=op,
+                translated_op=pipeline_ops.Mqtt_Publish(topic=topic, payload=op.message.data),
+            )
+
+        elif op.name is pipeline_ops.names.Iot_EnableFeature:
+            # Enabling a feature gets translated into an Mqtt subscribe operation
+            topic = self.subscription_map[op.feature_name]
+            self.continue_with_translated_op(
+                original_op=op, translated_op=pipeline_ops.Mqtt_Subscribe(topic=topic)
+            )
+
+        elif op.name is pipeline_ops.names.Iot_DisableFeature:
+            # Disabling a feature gets turned into an Mqtt unsubscribe operation
+            topic = self.subscription_map[op.feature_name]
+            self.continue_with_translated_op(
+                original_op=op, translated_op=pipeline_ops.Mqtt_Unsubscribe(topic=topic)
+            )
+
+        else:
+            # All other operations get passed down
+            self.continue_op(op)
+
+    def _set_topic_names(self, device_id, module_id):
+        """
+        Build topic names based on the device_id and module_id passed.
+        """
+        self.telemetry_topic = mqtt_topic_helper.get_telemetry_topic_for_publish(
+            device_id, module_id
+        )
+        self.subscription_map = {
```

Comment:
>subscription_map [](start = 13, length = 16)

I think this is a bad name. Map of what? Doesn't tell me that it has topics in it.